symbolic-ppdb/src/format/mod.rs

```diff
@@ -359,3 +367,85 @@ pub struct Document {
     pub name: String,
     pub(crate) lang: Language,
 }
+
+/// An iterator over Embedded Sources.
+#[derive(Debug, Clone)]
+pub struct EmbeddedSourceIterator<'object, 'data> {
+    ppdb: &'object PortablePdb<'data>,
+    inner_it: CustomDebugInformationIterator<'data>,
+}
+
+impl<'object, 'data> EmbeddedSourceIterator<'object, 'data> {
+    fn new(ppdb: &'object PortablePdb<'data>) -> Result<Self, FormatError> {
+        // https://github.com/dotnet/runtime/blob/main/docs/design/specs/PortablePdb-Metadata.md#embedded-source-c-and-vb-compilers
+        let embedded_sources_kind = Uuid::parse_str("0E8A571B-6926-466E-B4AD-8AB04611F5FE")
+            .map_err(|e| FormatError::new(FormatErrorKind::InvalidStringData, e))?;
+        let inner_it = CustomDebugInformationIterator::new(ppdb, embedded_sources_kind)?;
+        Ok(EmbeddedSourceIterator { ppdb, inner_it })
+    }
+}
+
+impl<'object, 'data> Iterator for EmbeddedSourceIterator<'object, 'data> {
+    type Item = Result<EmbeddedSource<'data>, FormatError>;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        self.inner_it.next().map(|inner| {
+            inner.and_then(|info| match info.tag {
+                // Verify we got the expected tag `Document` here.
+                metadata::CustomDebugInformationTag::Document => {
+                    let document = self.ppdb.get_document(info.value as usize)?;
+                    let blob = self.ppdb.get_blob(info.blob)?;
+                    Ok(EmbeddedSource { document, blob })
+                }
+                _ => Err(FormatErrorKind::InvalidCustomDebugInformationTag(info.tag as u32).into()),
+            })
+        })
+    }
+}
+
+/// Lazy Embedded Source file reader.
+#[derive(Debug, Clone)]
+pub struct EmbeddedSource<'data> {
+    document: Document,
+    blob: &'data [u8],
+}
+
+impl<'data, 'object> EmbeddedSource<'data> {
+    /// Returns the build-time path associated with this source file.
+    pub fn get_path(&'object self) -> Cow<'object, str> {
+        Cow::Borrowed(self.document.name.as_str())
+    }
+
+    /// Reads the source file contents from the Portable PDB.
+    pub fn get_contents(&self) -> Result<Cow<'data, [u8]>, FormatError> {
+        // The blob has the following structure: `Blob ::= format content`
+        // - format - int32 - Indicates how the content is serialized.
+        //     0 = raw bytes, uncompressed.
+        //     Positive value = compressed by deflate algorithm and value indicates uncompressed size.
+        //     Negative values reserved for future formats.
+        // - content - format-specific - The text of the document in the specified format. The length is implied by the length of the blob minus four bytes for the format.
+        let format = u32::from_ne_bytes(self.blob[0..4].try_into().unwrap());
```

Comment:
I’m a big fan of using `.get(0..4)` in these cases, as slice indexing can panic if you are out of bounds.

Reply:
Thanks for pointing the panic out. I went for `blob.split_at(4)`  with an explicit check beforehand instead of get(..) because I would have to write the error handling in three places in the same function.